lib/tests/modules/freeform-submodules.nix

```diff
@@ -0,0 +1,22 @@
+{ lib, options, ... }: with lib.types; {
+
+  options.fooDeclarations = lib.mkOption {
+    default = (options.free.type.getSubOptions [])._freeformOptions.foo.declarations;
+  };
+
+  options.free = lib.mkOption {
+    type = submodule {
+      config._module.freeformType = lib.mkMerge [
+        (attrsOf (submodule {
+          options.foo = lib.mkOption {};
+        }))
+        (attrsOf (submodule {
+          options.bar = lib.mkOption {};
+        }))
+      ];
+    };
+  };
+
+  config.free.xxx.foo = 10;
+  config.free.yyy.bar = 10;
+}
```

Comment:
Although this is the most compelling use case, I think it'd be wise to include one which tests an option with the "type" type as well. I don't know a real use case in nixpkgs, but this came pretty close:

https://github.com/NixOS/nixpkgs/blob/b5681a7a4096b81101626a0853c28bede0495490/nixos/modules/system/activation/top-level.nix#L194